### DISCUSSION TITLE
Not able to download audio with added Tone.js Effects

### DISCUSSION THREAD

**Discussion by imperium11** [naomiaro/waveform-playlist#198] [Q&A]
Hi,

I am having trouble downloading the edited audio tracks with added effects such as `reverb` from Tone.js. I followed the `basic-nextjs/pages/indexjs` file in order to get the download functionality to work (thank you Naomi) along with the example given on how to add `Reverb` - `AutoWah` etc utilizing the Tone.js library. 

The effects are working great but does not get added to the final downloaded audio file nor can i see the Tone.js effects if i log `playlist.getInfo()`

Thank you

**Comment by naomiaro** [comment]
Hmm thank-you for reporting 😄 

I didn't write the original exporter code, so I'll see if I missed something there 🐛

**Comment by naomiaro** [comment]
I've made a couple more releases to address these issues, but I've still come across cases when passing Tone.js an `OfflineAudioContext` for certain effects, they don't work properly and break the script. I'm not sure if this is a Tone.js limitation or not atm. I don't have problems with AutoWah, but for some other effects I've tried the output is either silent, or I get an error thrown. I'll spend a bit more time looking into this to see if I can do anything about it.

**Reply by imperium11** [reply]
Yes i had the same problem testing out different effects, usually its just silent. But all i need to use is like 3 effects Reverb, AutoWah and Distortion which all work on the tracks. So my download functionality was working before i updated to `4.3.0`. Now if i try to download, i get this error:

![Screenshot from 2022-02-23 10-12-33](https://user-images.githubusercontent.com/81498700/155400627-2c80e873-19b8-4b8a-b13a-c6015e813842.png)

As i mentioned before i followed the `basic-nextjs/pages/indexjs` download example and it was working. I'm guessing now its not working due to the added Tone.js effect?

**Reply by imperium11** [reply]
The tracks with the effects are working properly by the way. Just the download functionality. I'm wondering if this is a Tone.js bug

**Reply by naomiaro** [reply]
Tone.js has issues with certain effects and offline context that I'm seeing which is causing that error. At the moment I'm not sure if it's possible for some of these effects to work offline.

**Reply by naomiaro** [reply]
There's now a third argument to the effects function which will allow for different config based on whether effects are online or offline atm: https://github.com/naomiaro/waveform-playlist/blob/main/ghpages/js/effects.js#L44

**Reply by naomiaro** [reply]
Hmm just had a thought of a possible issue, will try to test it soon.

**Reply by imperium11** [reply]
Ok i changed my code to this and `Reverb` is working properly and downloadable

![Screenshot from 2022-02-23 11-19-47](https://user-images.githubusercontent.com/81498700/155410466-c98de625-5d0c-49cf-9427-53b4875346db.png)

**Reply by imperium11** [reply]
Also this is my download button (for the people that are wondering and come across this thread)

`<Button onClick={() => { ee.emit('startaudiorendering', 'wav'); }}>Download</Button>`

**Reply by naomiaro** [reply]
I found my issue with reverb being blank is the setup time the Tone.Reverb needs to initiate. I've added a way to push async setup needed for offline rendering to fix this issue atm. https://github.com/naomiaro/waveform-playlist/blob/main/ghpages/js/effects.js#L95

**Reply by naomiaro** [reply]
There's also a small demo showing how to set a custom toString on an effects function for better playlist info https://github.com/naomiaro/waveform-playlist/blob/main/ghpages/js/effects.js#L78

**Reply by imperium11** [reply]
thanks appreciate it. I created a drop down menu to select a track and add reverb on it (it takes the current playlist locates which track and adds the effects property with the reverb function) and before i put the updated playlist back on the editor, i just remove all tracks and load the updated playlist. It works fine but if you want to add `Reverb` to another track, it stringifies the first tracks `effects` property for some reason when you run playlist.getInfo().tracks which then the Reverb on the first track doesn't work anymore due to the fact that the effects property of the first track is a string and not a function. Do you happen to know why this could be?

**Reply by naomiaro** [reply]
I had called toString on the function objects, but looking at this now it doesn't make sense since nothing else in this output is a string yet. I'll change this to output just the functions. Thanks for helping with the testing of this 😄

**Reply by naomiaro** [reply]
Changed in v4.3.3

**Reply by imperium11** [reply]
ok going to try it now thanks

**Reply by imperium11** [reply]
Yup working great at the moment! I guess my last question to wrap everything up would be, when i try to download, why do i get several copies of the same downloaded audio track? Or maybe this just happens to me im not too sure.

I added 3 tracks to the editor added some effects etc then clicked download and it downloaded the same audio file 8-9 times.

![Screenshot from 2022-02-25 21-28-09](https://user-images.githubusercontent.com/81498700/155834366-09c1c842-cf18-4659-a6f1-3fb6c25db85f.png)

My code for the download functionality is:

![Screenshot from 2022-02-25 21-29-38](https://user-images.githubusercontent.com/81498700/155834404-fefc90f9-0cfd-404e-b27f-bf4002e22dd4.png)

And then i have a simple button: 
`<button onClick={() => { ee.emit('startaudiorendering', 'wav'); }}>Download</button>`

very similar to the nextjs example you made

Am i missing something here?

**Reply by naomiaro** [reply]
It's probably around where the events are placed on the event emitter. I changed the code a bit in my nextjs example https://github.com/naomiaro/waveform-playlist/blob/main/examples/basic-nextjs/pages/index.js to avoid events being registered multiple times by react re-renders.
